### PR TITLE
[hybrid] pp+dp support fp16 allreduce

### PR DESCRIPTION
**Pull request by wangxicoding**
### PR types

Performance optimization
### PR changes

Others
### Describe

Hybrid support fp16 allreduce. Usage:
``` python
import paddle.distributed.fleet as fleet

strategy = fleet.DistributedStrategy()
strategy.sharding = True
strategy.sharding_configs = {
    "sharding_degree": 1,
    "mp_degree": 1,
    "pp_degree": 2,
    "dp_degree": 2,
}
strategy.pipeline = True
strategy.pipeline_configs = {
    "schedule_mode": "1F1B",
    "micro_batch_size": 2,
    "accumulate_steps": 4,
}
strategy.amp = True
strategy.fp16_allreduce = True
```
#### Test
Test in 16node*8cards 32G V100, with Ernie3.0 model.

Model config:
| | value |
| -- | -- |
hidden size | 8192
num attention heads | 128
num hidden layers | 76
num sharing layers | 64
branch hidden size | 768
branch num attention heads | 16

Hybrid configs, with fused_allreduce 128MB:
dp | mp | pp | micro bsz | global bsz
-- | -- | -- | -- | --
2 | 8 | 8 | 2 | 256

**Performance:**
| fp16_allreduce | throughput(tokens/s) | improve |
| -- | -- | -- |
| false |  13394 | |
| true | 15224 | **13.6%** |

### CODE REVIEW

File: python/paddle/fluid/optimizer.py [hunk 1]
```diff
@@ -5044,14 +5044,20 @@ def _rename_gradient_var_name(self, block):
...
+                              pp_allreduce_in_optimize=False,
+                              fp16_allreduce=False):
         """
         Create a new merged gradient for each parameter and accumulate the
         corresponding gradient to it.
         """
         merged_gradient_names = []
         first_opt_op_idx = None
 
+        merged_suffix = '@MERGED@FP16' if fp16_allreduce else '@MERGED'
```
Comment (JZ-LIANG) [inline python/paddle/fluid/optimizer.py:5058]: should explain the suffix for gradname for later maintainer, we now have two many suffix for grad, immediately grad, accumulated grad, casted grad, etc

**Review by JZ-LIANG — APPROVED** [review]
LGTM

**Review by sandyhouse — APPROVED** [review]
LGTM

File: python/paddle/fluid/optimizer.py (hunk 1 above)
Comment (wangxicoding) [reply python/paddle/fluid/optimizer.py:5058]: ok, add in next PR

**Pull request merged**